pkg/provisioningv2/rke2/planner/encryptionkeyrotation.go

```diff
@@ -0,0 +1,450 @@
+package planner
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+	"time"
+
+	rkev1 "github.com/rancher/rancher/pkg/apis/rke.cattle.io/v1"
+	"github.com/rancher/rancher/pkg/apis/rke.cattle.io/v1/plan"
+	"github.com/rancher/rancher/pkg/controllers/provisioningv2/rke2"
+	"github.com/rancher/wrangler/pkg/generic"
+	"github.com/sirupsen/logrus"
+	capi "sigs.k8s.io/cluster-api/api/v1beta1"
+)
+
+const (
+	Start             = "start"
+	Prepare           = "prepare"
+	Rotate            = "rotate"
+	Reencrypt         = "reencrypt"
+	ReencryptRequest  = "reencrypt_request"
+	ReencryptActive   = "reencrypt_active"
+	ReencryptFinished = "reencrypt_finished"
+
+	SecretsEncryptApplyCommand  = "secrets-encrypt-apply-command"
+	SecretsEncryptStatusCommand = "secrets-encrypt-status-command"
+)
+
+func (p *Planner) RotateEncryptionKeys(cp *rkev1.RKEControlPlane, clusterPlan *plan.Plan) error {
+	if !shouldRotateKeys(cp) {
+		return nil
+	}
+
+	if shouldRestartEncryptionKeyRotation(cp) {
+		clusterPlan.Cluster.Annotations[capi.PausedAnnotation] = "true"
+		clusterPlan.Cluster.Spec.Paused = true
+		_, err := p.capiClient.Update(clusterPlan.Cluster)
+		if err != nil {
+			return p.enqueueAndSkip(cp)
+		}
+		return p.restartEncryptionKeyRotation(cp)
+	}
+	_, _, leader, err := p.findInitNode(cp, clusterPlan)
+	if err != nil {
+		return err
+	}
+	if leader == nil {
```

Comment:
An error message/log message for this would probably be helpful, otherwise we may run into situations where we don't know why the planner isn't progressing.